### DISCUSSION TITLE
Summing two streams for audio signal creation from multiple input signals

### DISCUSSION THREAD

**Discussion by neurologic** [bonsai-rx/bonsai#1007] [Q&A]
It seems like I am trying to do something doable, but I have not figured out how to obtain the final step.  
Essentially, I want to create an audio signal that is a sum of modulated sine waves. Ultimately, the amplitude of each sine wave that comprises the audio file will be modulated by the amplitude envelope of an EMG signal (if there is a way to do this within Bonsai that would be helpful, right now I am using BackyardBrains muscle shield that has a circuit within the amplifier that creates an envelope output).  
Anyway... To get the basic operations working I am first using **MouseMove** to modulate the amplitude of two sine waves (X position modulates one and Y position modulates the other). I have each stream of this workflow working (each stream is a separate modulated sine wave of a different frequency), but cannot figure out how to combine the two streams in a way that produces a signal that is a sum of both streams.  
I have tried a couple of things, but it feels like guessing a lot of the time. Most recently I tried **Zip** to put the two function generator outputs "together". I can create a .csv file from this zipped data successfully so that each column of the file is the signal from one of each function generator. However, I cannot figure out how to "add" the two channels in the zipped output.  I though **Sum(Dsp)** was a logical choice, but that gives me an error **"No method overload found for the given arguments"**. I also thought **Add(Expressions)** was promising, but get **"The binary operator Add is not defined for the types 'System.IntPtr' and 'System.IntPtr'**. Below is a screenshot of the workflow I have right now (and [here is the bonsai file](https://github.com/bonsai-rx/bonsai/files/9329548/audio.zip)).
![Capture](https://user-images.githubusercontent.com/4777333/184434370-50876fcf-bbfa-491d-942b-b6cbbbaed08d.PNG)

**Comment by bruno-f-cruz** [comment]
Hey!
I am willing to bet there is a way to do it using only `Mat` objects, but I dont know how. @glopesdev ?
As an alternative, I think you can simply convert them to an `Array` and do an element-wise sum. Check the attached file.
[sum_mat.zip](https://github.com/bonsai-rx/bonsai/files/9348760/sum_mat.zip)
Cheers,
B

**Reply by neurologic** [reply]
Thanks for your feedback and suggestions, Bruno.

Am I understanding that the amplitude scaling signal (ultimately the rate of the EMG signal) would now operate on the ```Double Value``` that ```Multiply Value``` subscribes to instead of directly to the ```Amplitude Value``` of the ```Function Generator```?

When I visualize the ```Add``` node in your suggested workflow, I do see a signal that is the sum of the two sine functions.

But,
I tried to attach an ```AudioPlayback``` node after the ```Add``` node and still got ***No method overload found for the given arguments***. Do you know what I might be misunderstanding? It seems like the Function Generator creates an F64 and the ```ConvertToArray``` node maintains that F64 depth, so I don't yet understand why the AudioPlayback node no longer accepts the output of the conversion operations.

If I strip down the workflow to: 
```Timer``` > ```Function Generator``` >``` repeat``` > ```AudioPlayback``` I am able to get playback of the audio (but at 441 I just hear a brief click each second. The sound is only continuous if the buffer is greater than 44100).
But, if I put the "```Take```> ```ConvertToArray``` > ```Merge```" sequence into the workflow, then the ```AudioPlayback``` fails.

What is the benefit of the ```Timer``` sequence start and ```Repeat``` sequence end? I originally just had the function generator as the start of the sequence and it seemed to create itself repeatedly by default. What is the downside of not using the timer?

**Comment by bruno-f-cruz** [comment]
Hey!

`Am I understanding that the amplitude scaling signal (ultimately the rate of the EMG signal) would now operate on the Double Value that Multiply Value subscribes to instead of directly to the Amplitude Value of the Function Generator?`

In theory both would work, I added the multiply to be more explicit and to operate on single values instead of whole buffer.

`I tried to attach an AudioPlayback node after the Add node and still got No method overload found for the given arguments. Do you know what I might be misunderstanding? It seems like the Function Generator creates an F64 and the ConvertToArray node maintains that F64 depth, so I don't yet understand why the AudioPlayback node no longer accepts the output of the conversion operations.`´

Ops sorry I only address the processing part. 'AudioPlayback' takes 'Mat' data types. You can convert the single elements back to a ´Mat´ using ´ConvertFromArray´ 

`What is the benefit of the Timer sequence start and Repeat sequence end? I originally just had the function generator as the start of the sequence and it seemed to create itself repeatedly by default. What is the downside of not using the timer?`

Timer -> ... -> Repeat just loops the function generator. The ´FunctionGenerator´ operator does not emmit events on its own. It needs an event to trigger the production of the buffer. In your case I see you are using Mouse events. Because I rather work with something a bit more controlled a used a Timer with a known DueTime.

I am attaching a modified version of the previous workflow. 
Hope it helps!
[sum_mat.zip](https://github.com/bonsai-rx/bonsai/files/9396841/sum_mat.zip)

**Comment by neurologic** [comment]
@bruno-f-cruz This is getting so much closer! I fiddled with the timer and buffer a bit. And I think I implemented a "maximum volume" rescaling after the addition of the streams together. This now produces a (mostly) continuous sound in real time and a great continuous audio file in which the real-time amplitude of each frequency is modulated by the x or y value of the mouse move (independently).
Perhaps you have advice on an issue I am having in the next big step.  

Now (attached) I am trying to use the amplitude of the signal on a channel of the analog input from a NIdaq (DAQmx) to scale the amplitude of the FunctionGenerator streams.  
BUT I have two problems (I think) that hopefully you or someone can help guide me on. 
1. When all of the nodes are enabled, I only get AudioPlayback, but there is no data streaming along the AnalogInput pathways. If I disable all of the Function Generator nodes, I do get AnalogInput data streaming through (I use the visualizers to see it). 
2. I figured out one way that seems to get the data from the OpenCV.Net.Mat (System.IntPtr) converted to System.Double for the Value property of the Multiply node. The Int node seemed to be the only one that worked (ConvertScale did not work). But the value of the integer does not make sense relative to the data. It is around 556638000 and seems to change with each buffer sweep (rather than reflecting an ongoing value of the AnalogInput Data). And it does not reflect the rescaled data (which is around 45000).  I tried ConverScale before Select Channels (thinking it needed to be converted to 32F to make sense to the Int node), but that did not help. 

Thank you so much for help and patience. I am trying to search through reactive expressions documentation to figure out these conversion issues, but I think I just don't know what to search for yet because I am not finding anything useful to my troubleshooting yet. 

[emg-music.zip](https://github.com/bonsai-rx/bonsai/files/9409092/emg-music.zip)
